File: 백준/Silver/10825. 국영수/국영수.cc
#include <iostream>
#include <algorithm>

using namespace std;

#define NM 100005

// 입력받을 자료형 선언
struct Elem{
    string name;
    int korean, eng, math;
    bool operator<(const Elem &rhs) const {
        if (korean != rhs.korean) return rhs.korean < korean;
        if (eng != rhs.eng) return eng < rhs.eng;
        if (math != rhs.math) return rhs.math < math;
        return name < rhs.name;
    }
}a[NM];

// 변수 선언
int N;

void input() {
    cin >> N;
    for (int i = 0; i < N; i++) {
        cin >> a[i].name >> a[i].korean >> a[i].eng >> a[i].math;
    }
}

void pro() {
    sort(a, a + N);
    for (int i = 0; i < N; i++) {
        cout << a[i].name << "\n"; 
    }
}

int main() {
    // 최적화
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    // 입력
    input();
    // 실행
    pro();

    return 0;
}

File: 백준/Silver/10825. 국영수/국영수.java
import java.io.*;
import java.util.*;

public class Main {
    static class Elem implements Comparable<Elem>{
        String name;
        int kor;
        int eng;
        int math;

        @Override
        public int compareTo(Elem other) {
            if (kor != other.kor)   return other.kor - kor;
            if (eng != other.eng)   return eng - other.eng;
            if (math != other.math) return other.math - math;
            return name.compareTo(other.name);
        }
        
    }

    static int N;
    static List<Elem> students = new ArrayList<>();

    static void input() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        for (int i = 0; i < N; i++) {
            students.add(new Elem());
            st = new StringTokenizer(br.readLine());

            students.get(i).name = st.nextToken();
            students.get(i).kor = Integer.parseInt(st.nextToken());
            students.get(i).eng = Integer.parseInt(st.nextToken());
            students.get(i).math = Integer.parseInt(st.nextToken());
        }
        br.close();
    }
    public static void main(String[] args) throws Exception{
        input();
        Collections.sort(students);
        for (int i = 0; i < N; i++) {
            System.out.println(students.get(i).name);
        }   
    }
}